Remove usage of !! from dotnet/runtime

stephentoub opened the pull request:
React to:
https://devblogs.microsoft.com/dotnet/csharp-11-preview-updates/#remove-parameter-null-checking-from-c-11
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-13.md#c-language-design-meeting-for-april-11th-2022

- Use ArgumentNullException.ThrowIfNull instead where possible.  It's only usable for projects that only target .NET 6+, and it can't be used in places like this(...) or base(...).
- In other cases, if the project already has a ThrowHelper, augment it for null as needed and use that.
- For most of the extensions projects, add a ThrowHelper.ThrowIfNull that replicates ArgumentNullException.ThrowIfNull.
- For everything else, just use "throw new".

stephentoub committed:
Remove usage of !! from dotnet/runtime

- Use ArgumentNullException.ThrowIfNull instead where possible.  It's only usable for projects that only target .NET 6+, and it can't be used in places like this(...) or base(...).
- In other cases, if the project already has a ThrowHelper, augment it for null as needed and use that.
- For most of the extensions projects, add a ThrowHelper.ThrowIfNull that replicates ArgumentNullException.ThrowIfNull.
- For everything else, just use "throw new".

GrabYourPitchforks commented:
Review based on acd12317c3d8d1b54ffd14a8a6624be8ab35059d

- _System.Formats.Asn1_ - OK
- _System.Formats.Cbor_ - OK
- _System.Memory_ - OK
- _System.Memory.Data_ - OK

GrabYourPitchforks commented:
The GitHub UI is misbehaving trying to leave feedback within the PR. But in __ClaimsIdentity.cs__: the 6.0 behavior is that for iterators, the _ArgumentNullException_ is thrown lazily, where in 7.0 it's thrown eagerly. I believe your 7.0 behavior is correct but wanted to surface it as a potential breaking change and make sure we were ok with it.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Windows.cs:
```diff
@@ -18,8 +18,10 @@ private ChainPal(SafeX509ChainHandle chain)
-                throw new ArgumentNullException(nameof(chainContext));
+            unsafe
+            {
+                ArgumentNullException.ThrowIfNull((void*)chainContext, nameof(chainContext));
```
Can we expose the `ArgumentNullException.ThrowIfNull(IntPtr)` overload publicly rather than require a pointer cast here?
Same comment in StorePal.\*.cs.

stephentoub replied:
(Hmm, I thought I'd actually undone my changes there... I guess they didn't stick.)

We _could_.  Especially with IntPtr and nint being one in the same, I'm not sure we want to new APIs that special-case IntPtr.Zero.  I'd expect moving forward any new APIs we provide that are pointer-based will actually take pointers, at which point we're mostly talking about doing so only for existing surface area. So, I'm not against another overload there, I'm just not sure it's worthwhile.

In the meantime, I'll revert the changes I'd made there as I'd intended to.

stephentoub commented:
>  But in ClaimsIdentity.cs and ClaimsPrincipal.cs: the 6.0 behavior is that for iterators, the ArgumentNullException is thrown lazily, where in 7.0 it's thrown eagerly. I believe your 7.0 behavior is correct but wanted to surface it as a potential breaking change and make sure we were ok with it.

Thanks, yes, I should have mentioned that.  In going through and manually reviewing every iterator and async method that had `!!`, I fixed a few to do the "right thing", which is to continue throwing the ArgumentNullException out of the method call rather than having it delayed until MoveNext.  There were these in Claims, a handful in Xml, and one or two others.  I think they fall under the bug-for-bug compat category and don't need to be called out as breaks unless we start to see lots of folks actually getting bitten by them.

stephentoub committed:
Address PR feedback

jeffhandley reviewed (commented):
`159 / 1412 files viewed`

I will continue tomorrow.

Tornhoof commented:
Just wondering, how did you undo the changes, it's a lot of changes and a large code base

1. Revert the changesets
2. Manual labor
3. Regex
4. Did you write a Roslyn codefix?

And if you did more than 1,2 how did you decide on that method?

stephentoub commented:
> how did you undo the changes, it's a lot of changes and a large code base

Revert wasn't ideal because:
- There were a lot of conflicts due to code merged since. 
- A fair amount of new use of `!!` has been added since.
- Rolling out `!!` fixed a decent number of (minor) bugs, e.g. wrong parameter name used, missing parameter name, missing tests, argument exceptions propagating asynchronously instead of synchronously out of iterators and async methods, etc. 
- We actually don't want what was there before but instead to use ThrowIfNull, as it's more concise and results in more efficient code, and `!!` is a good indicator of where it can/should be used.

My process:
1. Started by using VS to find expression-bodied members using `!!` and applying a refactoring to make them normal bodies. Most of these were in one LINQ-related file. 
2. Searched for async in combination with `!!` to manually do the "right thing" in terms of exceptions propagating synchronously. This found some bugs where we weren't doing it correctly prior to `!! `. Did the same searching for IEnumerable to find iterators, also finding a few bugs `!!` had fixed.
3. Searched for use of `!!` on signatures with this/base delegation to manually handle cases where the check needed to be before delegation. 
4. Wrote a Roslyn-based console app to find all `!!` on parameters, remove them, and add in an appropriate ThrowIfNull call. That handled 95% of the changes. 
5. Searched for ones my tool missed (I knew it would miss some but it was cheaper to do the rest manually than to make the tool bullet-proof), e.g. use on indexers, and fixed those manually. 
6. Built, fixed errors (primarily projects that couldn't use ThrowIfNull), rinsed and repeated until clean. 
7. Manually code-reviewed ~1400 files. Most of the issues here were around trivia my tool didn't handle appropriately, e.g. pragmas at the beginning of a method that got messed up.
8. Mentally praised `!!` and all that it accomplished for us in its few months with us, and put up the PR.

Tornhoof commented:
Thank you for the detailed answer.

stephentoub commented:
> 
<img width="211" alt="image" src="https://user-images.githubusercontent.com/2642209/164012822-0e2da60b-9878-43bc-9cc6-14bc7460b2b7.png">

I'll assume the downvotes are about C# 11 removing the `!!` feature rather than about the changes in this PR?  We need to remove use of `!!` if we want dotnet/runtime to continue to compile, which is, you know, important :smile:  If the downvotes are about the specifics of the changes here, I'm happy to hear specific feedback.

SteveDunn commented:
I've taken a look through a couple of files. One thing that has struck me is how nice it'd be if we could shorten (yes, I recognise the irony!) the null checking code, e.g, instead of:
```csharp
            ArgumentNullException.ThrowIfNull(element);
            ArgumentNullException.ThrowIfNull(attributeType);
```

we could have ...

```csharp
            ArgumentNullException.ThrowIfNull(element, attributeType);
```

I think this'd work if we added overloads, e.g.:

```csharp
public static void ThrowIfNull(
    [NotNull] object? argument1, 
    [NotNull] object? argument2, 
    [CallerArgumentExpression("argument1")] string? paramName1 = null,
    [CallerArgumentExpression("argument2")] string? paramName2 = null)
        {
            if (argument1 is null)
            {
                Throw(paramName1);
            }
            if (argument2 is null)
            {
                Throw(paramName2);
            }
        }
```

We could have up to 5 parameters for normal software (_and maybe a special 'Enterprise' version for Finance sector developers, with up to 30 parameters!_ ;) )

SteveDunn commented:
Actually, forget my idea above; [I implemented it](https://github.com/dotnet/runtime/pull/68234), but method overload resolution means it _likely_ will call the wrong overload.  Examples in the PR.

stephentoub committed:
Address PR feedback

stephentoub committed:
Remove false positives from searches

stephentoub commented on src/coreclr/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs:
```diff
@@ -343,10 +358,12 @@ protected virtual MethodInfo GetMethodImpl()
-        internal static Delegate CreateDelegateNoSecurityCheck(Type type!!, object? target, RuntimeMethodHandle method)
+        internal static Delegate CreateDelegateNoSecurityCheck(Type type, object? target, RuntimeMethodHandle method)
         {
             // Validate the parameters.
 
+            ArgumentNullException.ThrowIfNull(type);
```
Ok

stephentoub commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DefaultServiceProviderFactory.cs:
```diff
@@ -26,9 +26,9 @@ public DefaultServiceProviderFactory() : this(ServiceProviderOptions.Default)
+        public DefaultServiceProviderFactory(ServiceProviderOptions options)
         {
-            _options = options;
+            _options = options ?? throw new ArgumentNullException(nameof(options));
```
I don't think we should restrict using the ?? in this manner.

jeffhandley commented on src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/ContractUtils.cs:
```diff
@@ -46,28 +46,6 @@ public static void Requires([DoesNotReturnIf(false)] bool precondition, string p
-            }
-        }
-
         /// <summary>
```
I came here to see if this meant that we could get rid of `ContractUtils.RequiresNotNull`. 👍

jeffhandley reviewed (commented):
600 / 1416 files viewed. To be continued as soon as possible.

I found we're down to about 70 occurrences of an assignment with `?? throw new ArgumentNullException(nameof(...`. If we want to eliminate that pattern, we could do so in a follow-up PR pretty quickly, but I don't know if that was the aim or not.

stephentoub commented on src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:
```diff
@@ -18,8 +18,17 @@ public static class ChangeToken
-        public static IDisposable OnChange(Func<IChangeToken?> changeTokenProducer!!, Action changeTokenConsumer!!)
+        public static IDisposable OnChange(Func<IChangeToken?> changeTokenProducer, Action changeTokenConsumer)
         {
+            if (changeTokenProducer is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.changeTokenProducer);
+            }
+            if (changeTokenConsumer is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.changeTokenConsumer);
+            }
+
```
The pattern Microsoft.Extensions.Primitives is using here is slightly more efficient and is what a few libraries like corelib use.  It's using an enum value rather than a string to signal the name of the offending parameter, reducing the size of the call site even further.  We can't remove this pattern without measurements; it's possible it's overkill, it's possible it's meaningful.

jeffhandley replied:
Oh, that's enlightening. Thanks.

stephentoub commented on src/libraries/System.ComponentModel.Primitives/src/System/ComponentModel/InvalidEnumArgumentException.cs:
```diff
@@ -43,12 +43,13 @@ public InvalidEnumArgumentException(string? message, Exception? innerException)
                                 argumentName,
                                 invalidValue,
-                                enumClass.Name), argumentName)
+                                enumClass?.Name), argumentName)
         {
+            ArgumentNullException.ThrowIfNull(enumClass);
         }
```
> I'm curious why you chose not to throw from inside the call to base here (as was done in other occurrences of a similar scenario)

`throw new ArgumentNullException(nameof(enumClass))` results in larger code.  I only use it in this(...)/base(...) when I'm forced to in order to avoid null refs.  It's one of the nice things that `!!` provided, having the smaller code size of `ArgumentNullException.ThrowIfNull` but also being able to do that before the delegation.  We're losing that along with `!!`.

jeffhandley replied:
Got it; thanks.

stephentoub commented on src/libraries/System.Drawing.Common/tests/PenTests.cs:
```diff
@@ -905,8 +905,8 @@ public void MultiplyTransform_NullMatrix_ThrowsNullReferenceException()
             {
-                Assert.Throws<NullReferenceException>(() => pen.MultiplyTransform(null));
-                Assert.Throws<NullReferenceException>(() => pen.MultiplyTransform(null, MatrixOrder.Prepend));
+                Assert.Throws<ArgumentNullException>(() => pen.MultiplyTransform(null));
+                Assert.Throws<ArgumentNullException>(() => pen.MultiplyTransform(null, MatrixOrder.Prepend));
```
No

stephentoub commented on src/libraries/System.Collections.NonGeneric/src/System/Collections/Queue.cs:
```diff
@@ -66,7 +66,7 @@ public Queue(int capacity, float growFactor)
         // to get each of the elements.
         //
-        public Queue(ICollection col!!) : this(col.Count)
+        public Queue(ICollection col) : this(col?.Count ?? throw new ArgumentNullException(nameof(col)))
```
Yup.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting/src/Internal/ServiceFactoryAdapter.cs:
```diff
@@ -13,13 +13,16 @@ internal sealed class ServiceFactoryAdapter<TContainerBuilder> : IServiceFactory
+        public ServiceFactoryAdapter(IServiceProviderFactory<TContainerBuilder> serviceProviderFactory)
         {
-            _serviceProviderFactory = serviceProviderFactory;
+            _serviceProviderFactory = serviceProviderFactory ?? throw new ArgumentNullException(nameof(serviceProviderFactory));
```
Fixed

stephentoub commented:
> If we want to eliminate that pattern, we could do so in a follow-up PR pretty quickly, but I don't know if that was the aim or not.

My goal wasn't to remove them, but rather if a project already had a ThrowHelper, to use it consistently.

jeffhandley commented:
> Actually, forget my idea above; [I implemented it](https://github.com/dotnet/runtime/pull/68234), but method overload resolution means it _likely_ will call the wrong overload. Examples in the PR.

Thanks for trying that out, @SteveDunn. The idea was intriguing.

jeffhandley reviewed (approved):
I left one comment that calls out a subtle change in logic that we might want to back out for caution. Several other comments around code comments. And a couple formatting fixes/suggestions.

@stephentoub - your devotion to this was admirable. This was a major undertaking in both directions! And even though `!!` got removed, your pair of PRs around null argument handling resulted in better consistency as well as mitigations to some lurking NullReferenceException possibilities.

Thank you for the effort that went into this and for teaching me a few things during the review.

stephentoub committed:
Address PR feedback

stephentoub commented:
Thanks, @GrabYourPitchforks and @jeffhandley, for reviewing.
